price-feeder/oracle/provider/coinbase.go

```diff
@@ -0,0 +1,446 @@
+		return err
+	}
+
+	return nil
+}
+
+// subscribedPairsToSlice returns the map of subscribed pairs as a slice.
+func (p *CoinbaseProvider) subscribedPairsToSlice() []types.CurrencyPair {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	return mapPairsToSlice(p.subscribedPairs)
+}
+
+func (p *CoinbaseProvider) getTickerPrice(cp types.CurrencyPair) (TickerPrice, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	gp := currencyPairToCoinbasePair(cp)
+	if tickerPair, ok := p.tickers[gp]; ok {
+		return tickerPair.toTickerPrice()
+	} else {
+		return TickerPrice{}, fmt.Errorf("failed to get ticker price for %s", gp)
+	}
+}
+
+func (p *CoinbaseProvider) getTradePrices(key string) ([]CoinbaseTrade, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	trades, ok := p.trades[key]
+	if !ok {
+		return []CoinbaseTrade{}, fmt.Errorf("failed to get trades for %s", key)
+	}
+
+	return trades, nil
+}
+
+func (p *CoinbaseProvider) handleReceivedMessages(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewWSMessage):
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message.
+				p.logger.Err(err).Msg("could not read message")
+				if err := p.ping(); err != nil {
+					p.logger.Err(err).Msg("could not send ping")
+				}
+				continue
+			}
+
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.resetReconnectTimer()
+			p.messageReceived(messageType, bz)
+
+		case <-p.reconnectTimer.C: // reset by the pongHandler.
+			if err := p.reconnect(); err != nil {
+				p.logger.Err(err).Msg("error reconnecting")
+			}
+		}
+	}
+}
+
+func (p *CoinbaseProvider) messageReceived(messageType int, bz []byte) {
+	if messageType != websocket.TextMessage {
+		return
+	}
+
+	var coinbaseTrade CoinbaseTradeResponse
+	if err := json.Unmarshal(bz, &coinbaseTrade); err != nil {
+		p.logger.Debug().Msg("unable to unmarshal response")
+	} else {
+		if coinbaseTrade.Type == "error" {
+			var coinbaseErr CoinbaseErrResponse
+			if err := json.Unmarshal(bz, &coinbaseErr); err != nil {
+				p.logger.Debug().Msg("unable to unmarshal error response")
+			}
+			p.logger.Debug().Msg(coinbaseErr.Reason)
+		} else if coinbaseTrade.Type == "subscriptions" { // successful subscription message
+			return
+		} else if coinbaseTrade.Type == "ticker" {
+			var coinbaseTicker CoinbaseTicker
+			if err := json.Unmarshal(bz, &coinbaseTicker); err != nil {
+				p.logger.Debug().Msg("unable to unmarshal response")
+			}
+			p.setTickerPair(coinbaseTicker)
+		} else {
+			p.setTradePair(coinbaseTrade)
+		}
+	}
+}
+
+// timeToUnix converts a Time in format "2006-01-02T15:04:05.000000Z" to unix
+func (tr CoinbaseTradeResponse) timeToUnix() int64 {
+	t, err := time.Parse(timeLayout, tr.Time)
+	if err != nil {
+		return 0
+	}
+	// convert coinbase timestamp seconds -> milliseconds
+	return t.Unix() * int64(time.Second/time.Millisecond)
```

Comment:
`time.UnixMilli` might be better if you need exact milliseconds instead of nearest 1000